### ISSUE TITLE
why does compute method is different [EqualizedModulatedConv2d] from [EqualizedModulatedConvTranspose2d]

### ISSUE THREAD

**Issue by Johnson-yue** [yuuho/stylegans-pytorch#3]
the size of modulated_weight is [N, iC, oC, kH, kW]
the size of demod_norm is [N, oC]

https://github.com/yuuho/stylegans-pytorch/blob/d916bc0143767e815eb43a132c632300f600fb5b/packaged/run_pt_stylegan2.py#L157
This line is view demod_norm to [N, 1, oC, 1, 1]

but
https://github.com/yuuho/stylegans-pytorch/blob/d916bc0143767e815eb43a132c632300f600fb5b/packaged/run_pt_stylegan2.py#L197
this line is view demod_norm to [N, oC, 1, 1, 1]

this two function has a difficult result I guess

**Comment by yuuho** [comment]
Conv and transposed conv have differently shaped weight blocks.

[pytorch documentation](https://pytorch.org/docs/stable/nn.functional.html)

torch.nn.functional.conv2d(input, weight) [link](https://pytorch.org/docs/stable/nn.functional.html#torch.nn.functional.conv_transpose2d)
- input: (N,iC,H,W)
- weight: (oC,kH,kW)
- output: (N,oC,H,W)

torch.nn.functional.conv_transpose2d(input, weight) [link](https://pytorch.org/docs/stable/nn.functional.html#torch.nn.functional.conv2d)
- input: (N,iC,H,W)
- weight: (iC,oC,kH,kW)
- output: (N,oC,H,W)

I implemented according to this.
Each module holds the weight blocks in the required shape.

Thik of a single image. Module has oC weight blocks of size (iC, kH, kW).

In the modulation stage,
- calculate the modulation rate applied to each channel of the weight block.
- apply the same modulation rate to each weight block.

In the demodulation stage,
- each weight block is normalized to have a Euclidean norm of 1.

It can be simply written as follows.
```
# conv
assert self.weight.shape==(oC,iC,kH,kW)
modulated_weight = self.weight_scaler * self.weight.view(1,oC,iC,kH,kW) \
                    * mod_rates.view(N,1,iC,1,1) # (N,oC,iC,kH,kW)
demod_norm = 1 / ((modulated_weight**2).sum([2,3,4]) + 1e-8)**0.5 # (N, oC)
weight = modulated_weight * demod_norm.view(N, oC, 1, 1, 1) # (N,oC,iC,kH,kW)
```

However, in the case of transposed convolution, the dimensions of iC and oC are reversed, so
```
# transposed conv
assert self.weight.shape==(iC,oC,kH,kW)
modulated_weight = self.weight_scaler * self.weight.view(1,iC,oC,kH,kW) \
                    * mod_rates.view(N,iC,1,1,1) # (N,iC,oC,kH,kW)
demod_norm = 1 / ((modulated_weight**2).sum([1,3,4]) + 1e-8)**0.5 # (N, oC)
weight = modulated_weight * demod_norm.view(N, 1, oC, 1, 1) # (N,oC,iC,kH,kW)
```


The comment on transposed conv was wrong. Sorry.
https://github.com/yuuho/stylegans-pytorch/issues/2#issue-615130252
https://github.com/yuuho/stylegans-pytorch/blob/d916bc0143767e815eb43a132c632300f600fb5b/packaged/run_pt_stylegan2.py#L157

**Comment by Johnson-yue** [comment]
Thank you for your quick reply. I did n’t expect it. I did n’t check it in time due to the holiday. Sorry

I realized that I didn't understand the details of pytorch, here I will add two pictures to fully explain that you are right, but also to thank you for such a detailed answer, thank you very much

## torch.nn.functional.conv2d
![image](https://user-images.githubusercontent.com/10268274/81520185-2b4b4a80-9376-11ea-8ebf-d5429e8d79bb.png)

## torch.nn.functional.conv_transpose2d
![image](https://user-images.githubusercontent.com/10268274/81520242-56ce3500-9376-11ea-90e0-6ec36272869e.png)

Finally ，thank you add new comment for readable again
